pkg/prober/worker.go

```diff
@@ -58,35 +54,49 @@ func (w *Worker) stop() {
 
 // probe only supports TCP
 func (w *Worker) probe() error {
-	return w.tcpProber.Probe(w.address, w.timeout)
+	return w.tcpProber.Probe(w.Address, w.Timeout)
 }
 
 func (w *Worker) doProbe() {
+	// failure case
 	if err := w.probe(); err != nil {
-		logrus.Infof("probe error, %s, address: %s, timeout: %v", err.Error(), w.address, w.timeout)
 		w.successCounter = 0
 		w.failureCounter++
-	} else {
-		logrus.Infof("probe successful, address: %s, timeout: %v", w.address, w.timeout)
-		w.failureCounter = 0
-		w.successCounter++
-	}
-	if w.successCounter == w.successThreshold {
-		if !w.condition {
-			w.condition = true
-			w.conditionChan <- healthCondition{
-				workerUID: w.uid,
-				isHealth:  w.condition,
+		w.logSuccess = true
+		if w.failureCounter >= w.FailureThreshold {
+			// for continuous failure, only log error once in the controller life-cycle
+			if w.logFailure {
+				logrus.Infof("probe error uid:%s, address: %s, timeout: %v, error: %s", w.uid, w.Address, w.Timeout, err.Error())
+				w.logFailure = false
 			}
-		}
-	}
-	if w.failureCounter == w.failureThreshold {
-		if w.condition {
+			// notify anyway, the receiver may fail when processing
 			w.condition = false
 			w.conditionChan <- healthCondition{
-				workerUID: w.uid,
-				isHealth:  w.condition,
+				uid:       w.uid,
+				address:   w.Address,
+				isHealthy: w.condition,
 			}
+			w.failureCounter = 0
 		}
+		return
+	}
+
+	// successful case
+	w.failureCounter = 0
+	w.successCounter++
+	w.logFailure = true
+	if w.successCounter >= w.SuccessThreshold {
+		if w.logSuccess {
+			logrus.Infof("probe successful, uid:%s, address: %s, timeout: %v", w.uid, w.Address, w.Timeout)
+			w.logSuccess = false
+		}
+		// notify anyway, the receiver may fail when processing
+		w.condition = true
+		w.conditionChan <- healthCondition{
+			uid:       w.uid,
+			address:   w.Address,
+			isHealthy: w.condition,
+		}
+		w.successCounter = 0
```

Comment:
Probably, we don't need to reset `successCounter` here. We can modify if-condition like following, so we can avoid sending same healthCondition to the channel.

```go
if w.condition == true && w.successCounter >= w.SuccessThreshold {
```

Reply:
I feel in Jian's design, he wants to ensure the endpoint's ready condition remains as expected no matter if the value is `true` or `false`. This can prevent the condition from being modified by someone unexpectedly.

But I don't understand why resetting `w.successCounter` here and `w.failureCounter` at line 79.

Shouldn't the worker notify the manager for every probe after the counter surpasses the threshold?

Reply:
Yes, because the manager may encounter fail, and the manager can't enqueue itself. but the manager records last success /fail as a local var.

So the prober just reports success/fail to manager, the manager decides if continue populate to upstream.

Reply:
FYI:

https://github.com/harvester/load-balancer-harvester/blob/744c09f747bee515302bfafc4c6db26b918c2f39/pkg/lb/servicelb/manager.go#L86

https://github.com/harvester/load-balancer-harvester/blob/744c09f747bee515302bfafc4c6db26b918c2f39/pkg/prober/manager.go#L30

The existing code can only log an error when error happens, and it looks no chance to reconciler unless the pod is replaced.